### PR TITLE
Switch from ansi-wl-pprint to prettyprinter.

### PR DESCRIPTION
**Pull request by patrickt**
Given that the `ansi-wl-pprint` package is showing its age, I think it would be advantageous to switch to the `prettyprinter` package, which is establishing itself as the de facto prettyprinting library. This patch pulls in `prettyprinter` and, to support ANSI color sequences, `prettyprinter-ansi-terminal`. 

Some notes:
* I elected not to pull in `prettyprinter-compat-ansi-wl-pprint`, choosing instead to express `sgr` and `withHighlight` using `annotate`. There are a couple shims that I pulled over to avoid extensive refactors (`columns` and `renderPretty`).
* ~I’m a little suspicious of this `ANSIPretty` class; I might just split it into individual functions.~ I removed `Pretty` instances, using plain functions instead, which revealed that a few of said instances were never getting called.
* This is a breaking change, due to the `_errDoc` field being public. I 100% understand if you’d prefer to defer this change to version 3, or reject it altogether (though I would, of course, prefer not to have to maintain my own fork 😄 )

### CODE REVIEW

**Comment by patrickt** [comment]
Here’s a problem: `prettyprinter-ansi-terminal` requires `base >= 4.7`, which means that it won’t work on GHC 7.6 or 7.4

I’m not sure what to do about this. Though the simplest course of action would be to drop support for ancient GHCs (7.6 came out in 2013, which makes it practically geologically old), I doubt that this will be a popular option. Here are some alternatives:
* I could try to get a patch into `prettyprinter-ansi-terminal` to relax the `base` version, but that might not be acceptable to its maintainer.
* I could copy the required code from `prettyprinter-ansi-terminal` wholesale, which would work but entails ugly duplication.
* I could abandon `prettyprinter-ansi-terminal` altogether and use `Doc SGR`, then write my own implementations of `renderIO` and `renderPretty`. This has the potential to introduce bugs.
* I could abandon this patch.

I’m not sure of the best course of action. If @phadej or any other interested parties have suggestions, I’m all ears.

**Comment by patrickt** [comment]
For context, the initial discussion about moving to `prettyprinter` is here https://github.com/ekmett/trifecta/pull/83#issuecomment-454902223

**Comment by nwf** [comment]
I, personally, am fine with abandoning GHC 7.6 and earlier, FWIW.  Perhaps another approach, though, would be to drop ANSI terminal support with `base < 4.7` and leave the rest of the package functional?

**Comment by phadej** [comment]
First of all: thanks!

Hackage is down, so I cannot easily check, but looks like latest `prettyprinter` (I checked `1.2.1` and `1` versions) itself has `base >= 4.7` (GHC-7.8+)
IMHO, we can drop older GHC support. If someone complains, I'd be happy to help, *then* (I don't think that `prettyprinter` uses anything GHC-7.8 brought to the table, I might be wrong).

Let me look through the patch next

File: src/Text/Trifecta/Pretty.hs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+#if __GLASGOW_HASKELL__ < 710
+import Data.Monoid
+#endif
+
+import Data.Text.Prettyprint.Doc
+import Data.Text.Prettyprint.Doc.Render.Terminal
+import Data.Text.Prettyprint.Doc.Render.Terminal.Internal (ansiBold, ansiUnderlining)
+
+class ANSIPretty a where
```
Comment (phadej) [inline src/Text/Trifecta/Pretty.hs]: I see. When I thought about the migration my first idea that this is fishy. I'd rather have `prettyDelta`, `prettyHighlightedRope` etc. There are just a few, and AFAIK we never abstract over them (i.e. we don't have `AnsiPretty a => ...` functions).

**Comment by phadej** [comment]
I'm not sure about this anymore.

Maybe we should just make a trifecta release and postpone this `prettyprinter` business until `optparse-applicative` either gets the support or completely rejects the idea.

**Comment by RyanGlScott** [comment]
I must be missing something here. `trifecta` doesn't depend on `optparse-applicative`, right? Why should the latter hold up the former?

**Comment by phadej** [comment]
Until optparse-applicative decides the ansi-wl-print is still around, so changing to prettyprinter in trifecta won’t remove ansi-wl-pprint from install plans.

One could deprecate ansi-wl-pprint on Hackage to push people to change, but
- it didn’t really happened with mustache libraries (i.e. deprecation had no noticeable effect)
- atm ansi-wl-pprint is actully in a good shape, and its maintenance is not a burden

Or simply: I lost interest in pushing prettyprinter adoption forward, and ansi-wl-pprint is good enough.

**Comment by RyanGlScott** [comment]
I'm still lost. This PR is all about `trifecta`, which has no relationship whatsoever with the `optparse-applicative` library. Even if `optparse-applicative` doesn't switch over to `prettyprinter`, I don't think that's a valid reason to stall this PR—it's a completely independent effort!

In other words, I'm wondering why we can't just approach this one step at a time. I admire your desire to replace all uses of `ansi-wl-pprint` with `prettyprinter` in popular libraries, but at the same time, I don't think we should let perfect be the enemy of good. As far as I can tell, switching `trifecta` over to `prettyprinter` is quite manageable today, so if that is the case, let's merge this and worry about `optparse-applicative` another day.

**Comment by RyanGlScott** [comment]
Disclaimer: my comment in https://github.com/ekmett/trifecta/pull/84#issuecomment-493449736 was written under the assumption that `prettyprinter` is already suitable for use in `trifecta`. If that isn't the case for some reason, then that would be worth taking into account.

**Comment by ekmett** [comment]
I think I'm inclined to start pulling the trigger on moving all of my packages that depend on `ansi-wl-pprint` over to `pretty-printer`.

I'm also willing to put a hard floor under whatever the new version is of something like GHC 7.10 or so to clean up cruft and start modernizing this library.

Otherwise it'll just bitrot forever.

**Pull request merged**